pgxn/neon/libpagestore.c

```diff
@@ -303,119 +329,275 @@ get_shard_number(BufferTag *tag)
 	return hash % n_shards;
 }
 
+/*
+ * Connect to a pageserver, or continue to try to connect if we're yet to
+ * complete the connection (e.g. due to receiving an earlier cancellation
+ * during connection start).
+ * Returns true if successfully connected; false if the connection failed.
+ * 
+ * Throws errors in unrecoverable situations, or when this backend's query
+ * is canceled.
+ */
 static bool
 pageserver_connect(shardno_t shard_no, int elevel)
 {
-	char	   *query;
-	int			ret;
-	const char *keywords[3];
-	const char *values[3];
-	int			n;
-	PGconn	   *conn;
-	WaitEventSet *wes;
+	PageServer *shard = &page_servers[shard_no];
+	PGconn	   *conn = shard->conn;
+	WaitEventSet *wes_read = shard->wes_read;
+	WaitEventSet *wes_write = shard->wes_write;
```

Comment:
I recommend removing these local variables and always using `shard->conn`, `shard->wes_read`, `shard->wes_write`. That way they cannot be out of sync. I don't see such bugs currently, but it's an easy mistake to forget to update the local variable when the struct member is updated.